File: python/ticket_office/ticket_office/infra/rest_client.py
from httpx import Client as HttpClient

from ticket_office.domain.client import Client
from ticket_office.domain.seat import Seat
from ticket_office.domain.train import Train


class ReservationFailed(Exception):
    def __init__(self, status_code, text):
        super().__init__(status_code, text)


class RestClient(Client):
    def __init__(self):
        self._http_client = HttpClient()

    def get_booking_reference(self):
        return self._http_client.get("http://localhost:8082/booking_reference").text

    def get_train(self, train_id):
        train_data = self._http_client.get(
            f"http://localhost:8081/data_for_train/" + train_id
        ).json()
        seats = []
        for json_seat in train_data["seats"].values():
            number = json_seat["seat_number"]
            coach = json_seat["coach"]
            booking_reference = json_seat["booking_reference"]
            if booking_reference:
                seat = Seat.booked(
                    number=number, coach=coach, booking_reference=booking_reference
                )
            else:
                seat = Seat.free(number=number, coach=coach)
            seats.append(seat)

        return Train(seats)

    def make_reservation(self, *, train_id, booking_reference, seat_ids):
        reservation_payload = {
            "train_id": train_id,
            "booking_reference": booking_reference,
            "seats": seat_ids,
        }
        response = self._http_client.post(
            "http://localhost:8081/reserve",
            json=reservation_payload,
        )
        if response.status_code != 200:
            raise ReservationFailed(response.status_code, response.text)

File: ts/ticket-office/infra/RestClient.ts
import Booking from "../domain/Booking"
import BookingReferenceSource from "../domain/BookingReferenceSource"
import Seat, { CoachId, SeatId, SeatNumber } from "../domain/Seat"
import Train from "../domain/Train"
import TrainRepository from "../domain/TrainRepository"

export default class RestClient implements BookingReferenceSource, TrainRepository {
  async getTrain(trainId: string): Promise<Train> {
    const response = await fetch(`http://localhost:8081/data_for_train/${trainId}`)
    const json = await response.json()
    return parseTrain(json)
  }

  async applyBooking(booking: Booking): Promise<void> {
    const { bookingReference, seatIds, trainId } = booking
    const reservation = {
      booking_reference: bookingReference,
      seats: seatIds,
      train_id: trainId,
    }
    const response = await fetch(`http://localhost:8081/reserve`, {
      method: 'POST',
      body: JSON.stringify(reservation),
      headers: { 'Content-Type': 'application/json' }
    })
    const status = response.status
    if (status != 200) {
      const text = await response.text()
      throw new Error(`Got code ${status} when making reservation\n${text}`)
    }
  }

  async getNewBookingReference(): Promise<string> {
    const response = await fetch("http://localhost:8082/booking_reference")
    return response.text()
  }

}

type JsonSeat = {
  coach: string,
  seat_number: string,
  booking_reference: string,
}

type JsonTrain = {
  "seats": Map<string, JsonSeat>,
}

const parseTrain = (json: JsonTrain): Train => {
  const seats: Seat[] = []
  const jsonSeats = Object.values(json.seats)
  jsonSeats.forEach(jsonSeat => {
    const { coach, seat_number, booking_reference } = jsonSeat
    const coachId = new CoachId(coach)
    const seatNumber = new SeatNumber(seat_number)
    const seatId = new SeatId(seatNumber, coachId)
    const seat = new Seat(seatId, booking_reference)
    seats.push(seat)
  })
  return new Train(seats)
}